chainer/functions/connection/multihead_attention.py

```diff
@@ -0,0 +1,298 @@
+import typing as tp  # NOQA
+
+import chainer
+from chainer.functions.activation import softmax
+from chainer.functions.array import concat
+from chainer.functions.array import expand_dims
+from chainer.functions.array import split_axis
+from chainer.functions.array import repeat
+from chainer.functions.array import reshape
+from chainer.functions.array import transpose
+from chainer.functions.array import where
+from chainer.functions.math import average
+from chainer.functions.math import matmul
+from chainer.functions.noise import dropout
+from chainer.functions.connection import linear
+from chainer import types
+from chainer import variable
+
+
+InputType = tp.Union[variable.Variable, types.NdArray]
+
+
+def generate_zeros(device, shape, dtype):
+    with chainer.using_device(device):
+        return device.xp.zeros(shape=shape, dtype=dtype)
+
+
+def generate_ones(device, shape, dtype):
+    with chainer.using_device(device):
+        return device.xp.ones(shape=shape, dtype=dtype)
+
+
+def multihead_attention(
+    n_head,                   # type: int
+    embedding_size,           # type: int
+    query,                    # type: InputType
+    key,                      # type: InputType
+    value,                    # type: InputType
+    proj_in_W,                # type: tp.Union[variable.Variable, tp.Tuple[variable.Variable, variable.Variable, variable.Variable]]  # NOQA
+    proj_in_b,                # type: tp.Optional[variable.Variable]
+    bias_k,                   # type: variable.Variable
+    bias_v,                   # type: variable.Variable
+    proj_out_W,               # type: variable.Variable
+    proj_out_b,               # type: variable.Variable
+    add_zero_attn=False,      # type: bool
+    attention_dropout=0,      # type: float
+    post_dropout=0,           # type: float
+    key_padding_mask=None,    # type: tp.Optional[InputType]
+    attn_mask=None,           # type: tp.Optional[InputType]
+    dot_product_scaler=None,  # type: tp.Optional[float]
+    softmax_scaler=1.0,       # type: float
+    return_weights=True       # type: bool
+):
+    # type: (...) -> tp.Tuple[variable.Variable, variable.Variable]  # NOQA
+    """Multi-head Attention forward function.
+
+    Args:
+        query (:class:`~chainer.Variable` or :ref:`ndarray`):
+            A batch of query vectors whose shape is :math:`(L, B, E)` where
+            :math:`L` is the target sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
```

Comment:
Could you explain why the batch dimension isn't placed first?

Reply:
@hvy Does the above explanation make sense to you?